File: app/src/main/java/me/ykrank/s1next/view/adapter/delegate/ThreadAdapterDelegate.java
package me.ykrank.s1next.view.adapter.delegate;

import android.content.Context;
import android.databinding.DataBindingUtil;
import android.support.annotation.NonNull;
import android.support.v7.widget.RecyclerView;
import android.view.ViewGroup;

import javax.inject.Inject;

import me.ykrank.s1next.App;
import me.ykrank.s1next.R;
import me.ykrank.s1next.data.api.model.Thread;
import me.ykrank.s1next.data.pref.ThemeManager;
import me.ykrank.s1next.databinding.ItemThreadBinding;
import me.ykrank.s1next.viewmodel.ThreadViewModel;
import me.ykrank.s1next.viewmodel.UserViewModel;

public final class ThreadAdapterDelegate extends BaseAdapterDelegate<Thread, ThreadAdapterDelegate.BindingViewHolder> {

    @Inject
    UserViewModel mUserViewModel;

    @Inject
    ThemeManager mThemeManager;

    public ThreadAdapterDelegate(Context context) {
        super(context);

        App.getPrefComponent(context).inject(this);
    }

    @NonNull
    @Override
    protected Class<Thread> getTClass() {
        return Thread.class;
    }

    @NonNull
    @Override
    public RecyclerView.ViewHolder onCreateViewHolder(ViewGroup parent) {
        ItemThreadBinding binding = DataBindingUtil.inflate(mLayoutInflater, R.layout.item_thread,
                parent, false);
        // we do not use view model for ThemeManager
        // because theme changes only when Activity recreated
        binding.setUserViewModel(mUserViewModel);
        binding.setThemeManager(mThemeManager);
        binding.setThreadViewModel(new ThreadViewModel());

        return new BindingViewHolder(binding);
    }

    @Override
    public void onBindViewHolderData(Thread thread, int position, @NonNull BindingViewHolder holder) {
        ItemThreadBinding binding = holder.itemThreadBinding;
        binding.getThreadViewModel().thread.set(thread);
        binding.getThreadViewModel().setSubscription();
        binding.executePendingBindings();
    }

    static final class BindingViewHolder extends RecyclerView.ViewHolder {

        private final ItemThreadBinding itemThreadBinding;

        public BindingViewHolder(ItemThreadBinding itemThreadBinding) {
            super(itemThreadBinding.getRoot());

            this.itemThreadBinding = itemThreadBinding;
        }
    }
}

File: app/src/main/java/me/ykrank/s1next/view/adapter/delegate/ThreadAdapterDelegate.kt
package me.ykrank.s1next.view.adapter.delegate

import android.content.Context
import android.view.ViewGroup
import androidx.databinding.DataBindingUtil
import com.github.ykrank.androidtools.ui.adapter.simple.SimpleRecycleViewHolder
import me.ykrank.s1next.App
import me.ykrank.s1next.R
import me.ykrank.s1next.data.api.model.Thread
import me.ykrank.s1next.data.pref.ReadPreferencesManager
import me.ykrank.s1next.data.pref.ThemeManager
import me.ykrank.s1next.databinding.ItemThreadBinding
import me.ykrank.s1next.viewmodel.ThreadViewModel
import me.ykrank.s1next.viewmodel.UserViewModel
import javax.inject.Inject

class ThreadAdapterDelegate(context: Context, private val forumId: String) :
        BaseAdapterDelegate<Thread, SimpleRecycleViewHolder<ItemThreadBinding>>(context, Thread::class.java) {

    @Inject
    internal lateinit var mUserViewModel: UserViewModel

    @Inject
    internal lateinit var mThemeManager: ThemeManager

    @Inject
    internal lateinit var mReadPreferencesManager: ReadPreferencesManager

    init {
        App.appComponent.inject(this)
    }

    public override fun onCreateViewHolder(parent: ViewGroup): androidx.recyclerview.widget.RecyclerView.ViewHolder {
        val binding = DataBindingUtil.inflate<ItemThreadBinding>(mLayoutInflater, R.layout.item_thread,
                parent, false)
        // we do not use view model for ThemeManager
        // because theme changes only when Activity recreated
        binding.userViewModel = mUserViewModel
        binding.themeManager = mThemeManager
        binding.forumId = forumId
        binding.model = ThreadViewModel()

        val threadPadding = mReadPreferencesManager.threadPadding
        if (threadPadding != null && threadPadding > 0) {
            val paddingPx = threadPadding * parent.context.resources.displayMetrics.scaledDensity.toInt()
            binding.tvThread.setPadding(binding.tvThread.paddingLeft, paddingPx, binding.tvThread.paddingLeft, paddingPx)
        }

        return SimpleRecycleViewHolder<ItemThreadBinding>(binding)
    }

    override fun onBindViewHolderData(t: Thread, position: Int, holder: SimpleRecycleViewHolder<ItemThreadBinding>, payloads: List<Any>) {
        val binding = holder.binding
        binding.model?.thread?.set(t)
    }

}
